content/blog/foreign-key-constraints.md

```diff
@@ -0,0 +1,198 @@
+
+## PostgreSQL
+
+Postgres databases capture schema metadata within catalogs. These catalogs are just regular tables which we can query as normal. We're going to use three different tables to extract constraint metadata in a readable form:
+
+* `pg_constraint` - contains data about constraints of different types (check, primary key, unique, foreign key and exclusion) for all tables
+* `pg_class` - contains data about tables in the database (it also has information about other relations e.g. sequences, views etc. but we're only interested in the tables here)
+* `pg_attribute` - contains information about table columns with one entry for every column of every table in the database
+
+From these the fields we are most interested in are:
+
+### `pg_constraint`
+* `oid` - row identifier
+* `conname` - constraint name
+* `contype` - constraint type, `f` = foreign key constraint, `p` = primary key constraint, etc.
+* `conkey` - List of constrained columns
+* `conrelid` - `oid` of the table this constraint is on
+* `confkey` - List of referenced columns for foreign key
+* `confrelid` - `oid` of the referenced table for a foreign key
+
+### `pg_class`
+* `oid` - row identifier
+* `relname` - the name of the table
+
+### `pg_attribute`
+* `attname` - the column name
+* `attrelid` - the oid the table belongs to
+* `attnum` - the number of the column
+
+Let's set up an example DB with some foreign key constraints. We create 3 tables with 3 foreign key constraints:
+
+```
+CREATE TABLE region (
+  name TEXT PRIMARY KEY,
+  parent TEXT,
+  FOREIGN KEY (parent) REFERENCES region(name)
+);
+CREATE TABLE street (
+  name TEXT PRIMARY KEY
+);
+CREATE TABLE address (
+  street TEXT,
+  region TEXT,
+  FOREIGN KEY (street) REFERENCES street(name),
+  FOREIGN KEY (region) REFERENCES region(name)
+);
+```
+
+<img src="/img/example_db.png" alt="png of database schema"/>
+
+We want to develop a query to pull out the information in a readable form from the above tables. This involves getting the list of constraints from `pg_constraint` then joining `pg_class` and `pg_attribute` on `oid` to pull out the name of the table and column. The difficulty is that conkey and confkey fields are both lists so we need to use `UNNEST` to expand these to get a single row for each constraint. 
+
+We can use common table expressions (`WITH` clauses) to do this. We need two queries to get the `conkey` and `confkey` with the corresponding `oid`:
+```
+SELECT oid, unnest(confkey) as confkey
+FROM pg_constraint;
+```
+
+```
+SELECT oid, unnest(conkey) as conkey
+FROM pg_constraint;
+```
+
+Then we `LEFT JOIN` to these expressions in the main query to get a row for every combination of constrained column to referenced column to get the full query:
+
+```
+WITH unnested_confkey AS (
```

Comment:
do you want to show how to simplify this with the lateral join trick?

Reply:
I was thinking about leaving that off. It feels like a big aside to talk introduce lateral join, explain what it does, mention how we use it with `WITH ORDINALITY`. And to be honest I find this version easier to understand than the lateral join version - but that may well be because I wrote this one.

I was thinking the separate blog post about lateral join and I could reference this one as an application but happy to add if you think it will benefit this post.